extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerRunTimeConfig.java

```diff
@@ -0,0 +1,33 @@
+package io.quarkus.oidc.db.token.state.manager.runtime;
+
+import static io.quarkus.runtime.annotations.ConfigPhase.RUN_TIME;
+
+import java.time.Duration;
+
+import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+
+@ConfigMapping(prefix = "quarkus.oidc-db-token-state-manager")
+@ConfigRoot(phase = RUN_TIME)
+public interface OidcDbTokenStateManagerRunTimeConfig {
+
+    /**
+     * Whether Quarkus should delete expired token states.
+     * You can disable this feature when you want only one of your service instances to delete expired tokens.
+     */
+    @WithDefault("true")
+    boolean deleteExpired();
```

Comment:
I'm not sure that this property is required. If the cookie representing this DB data itself has expired the browser will remove it and it won't reach Quarkus and `TokenStateManager#remove` will not be even called. I.e, if this property is `false`, DB will slowly keep growing

Reply:
> I'm not sure that this property is required. If the cookie representing this DB data itself has expired the browser will remove it and it won't reach Quarkus and `TokenStateManager#remove` will not be even called. I.e, if this property is `false`, DB will slowly keep growing

Consider following situations:

- in OpenShift you are most likely to run your service in more than one track. E.g. you scale to 2-3, whatever works for you. but it doesn't make sense for every pod to try to delete expired tokens, there is no additional value, hence you only enable it on one instance (it can be automated, you will only set this env var on one of pods via deployment config)
- you definitely won't to delete tokens that have expired and browser removed them, but would stick in our database in cases when TokenStateManager#deleteToken wasn't called. it could be potential point of attack if database keep growing if the attacker was authenticated, he would just drop the old token and authenticated again, and again, until DoS succeed

That's way it is there.

Reply:
BTW I can imagine other, though less likely scenarios when you want to stop delete. You can have service with cluster schedulers and do periodic stuff only in there, so users will delete expired tokens themselves. Or someone may prefer trigger (it wouldn't be very performant though).

I suppose your concern is rather about keeping as few configuration options as possible? I don't have strong opinion there, but for arguments I provided, I vote to keep this flag.

Reply:
> My point is, if this property is ever set to false then how the expired tokens will be removed, given that we know that if the only link to the data in the form of the cookie has been deleted by the browser, ex, when the user cleared the cache or the cookie has expired ?

Yeah, I'm having trouble to find words we would both understand. Hope you won't get offended if I try to simplify it:

- create quarkus project hello-app
- deploy it to openshift -> hello-app-svc
- scale to 2 pods
- after 8 h, both pods are deleting expired tokens; I'm not worried about race, but it doesn't make sense. 

now instead, I create 2 deployment configs, both will scale to one (2x1=2):
-  with env var that disables deletion
- without env var that will delete tokens, this deployment config will never scale to zero because you always want to keep such service up

I think that is the very reason why cluster schedulers exists, so that you don't execute in all app instances, but just one of them. And so, if someone has already app let say with Quartz job, he can also disable deletion and instead run delete statement himself. (I didn't document this last paragraph, for I think it's unlikely, it would just add noise to docs).

Reply:
@michalvavrik 

>after 8 h, both pods are deleting expired tokens

I wonder if if it is where the problem is. I.e, my understanding we don't introduce this feature as Kubernetes native feature - for such multi-pods set ups, well, let users have it in the encrypted cookies, the problem is solved. We offer it for simple cases, no high throughput flows.

Or if the users insist on using it in Kubernetes, then like you said, both pods will send delete requests to the DB - is there a problem there apart it is suboptimal ? At least it is guaranteed to happen, as opposed to depending on some human intervention to prepare the scheduler specifically to manage the deletion of the token records. 

I'd not mind, at least for now, dropping this property and reintroduce it, now that I know why you considered it, when someone actually tries to use the feature in the cloud in a multi-pod setup, and considers it being suboptimal.

Reply:
If you prefer to keep then IMHO to make it very clear in the docs that users should not disable it in most cases, only consider doing in a set up you have described with the scheduler being used instead. 
That should be ok then I guess, it will prevent users from trying do some optimizations which will lead to the leaks.
I'll let you decide, I'll support

Reply:
> Or if the users insist on using it in Kubernetes, then like you said, both pods will send delete requests to the DB - is there a problem there apart it is suboptimal ? At least it is guaranteed to happen, as opposed to depending on some human intervention to prepare the scheduler specifically to manage the deletion of the token records.

No, it's not an issue, really. You convinced me, we can drop it. The documentation part, also makes sense, it would need to be better documented. **I am going to drop the flag.**

> for such multi-pods set ups, well, let users have it in the encrypted cookies, the problem is solved. We offer it for simple cases, no high throughput flows.

I never mentioned it before, for you keep mentioning simple apps and know this topic better than I do. But here are my 2 cents: Sending encrypted token is safe as it is highly unlikely that user encryption is going to be broken or secret is going to be exposed. But I can imagine someone will still don't like it, for sending only db pointer is safer. And I think should we store tokens in cluster cache like REDIS, it's going to be fast and cloud native as well (e.g. in OpenShift you have REDIS operator that is easy peasy to setup), e.g. using Quarkus Redis Client, and it would be suitable for high throughput flows. Anyway, I get there must be users demand first.

I think we should close this thread. Thanks for input.

Reply:
Done.

Reply:
Thanks @michalvavrik We can certainly review this area of dealing with the expired tokens once users start asking